pkg/streamingpromql/types/ring_buffer_test.go

```diff
@@ -563,3 +563,363 @@ func setupRingBufferTestingPools(t *testing.T) {
 		putHPointSliceForRingBuffer = originalPutHPointSlice
 	})
 }
+
+func TestFPointRingBufferInsertHeadPointsTimestampCheck(t *testing.T) {
+	buff := NewFPointRingBuffer(limiter.NewMemoryConsumptionTracker(context.Background(), 0, nil, ""))
+	require.NoError(t, buff.InsertHeadPoint(promql.FPoint{T: 10, F: 20}))
+	require.ErrorContains(t, buff.InsertHeadPoint(promql.FPoint{T: 20, F: 30}), "unable to insert point to the head of the buffer - new point must have a timestamp less then the current first point timestamp")
+}
+
+func TestFPointRingBufferDiscardPointsAtOrBefore(t *testing.T) {
+	testCases := map[string]struct {
+		timestamp int64
+		buff      []promql.FPoint
+		expected  []promql.FPoint
+	}{
+		"empty buff": {
+			timestamp: 0,
+			buff:      []promql.FPoint{},
+			expected:  nil,
+		},
+		"single point - remove on boundary": {
+			timestamp: 10,
+			buff:      []promql.FPoint{{T: 10, F: 20}},
+			expected:  nil,
+		},
+
+		"single point - remove not on boundary": {
+			timestamp: 12,
+			buff:      []promql.FPoint{{T: 10, F: 20}},
+			expected:  nil,
+		},
+
+		"single point - remove nothing": {
+			timestamp: 9,
+			buff:      []promql.FPoint{{T: 12, F: 20}},
+			expected:  []promql.FPoint{{T: 12, F: 20}},
+		},
+
+		"multiple points - remove all": {
+			timestamp: 100,
+			buff:      []promql.FPoint{{T: 10, F: 20}, {T: 20, F: 20}, {T: 30, F: 20}, {T: 40, F: 20}, {T: 50, F: 20}, {T: 60, F: 20}, {T: 70, F: 20}, {T: 80, F: 20}},
+			expected:  nil,
+		},
+
+		"multiple points - remove some": {
+			timestamp: 35,
+			buff:      []promql.FPoint{{T: 10, F: 20}, {T: 20, F: 20}, {T: 30, F: 20}, {T: 40, F: 20}, {T: 50, F: 20}, {T: 60, F: 20}, {T: 70, F: 20}, {T: 80, F: 20}},
+			expected:  []promql.FPoint{{T: 40, F: 20}, {T: 50, F: 20}, {T: 60, F: 20}, {T: 70, F: 20}, {T: 80, F: 20}},
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			buff := NewFPointRingBuffer(limiter.NewMemoryConsumptionTracker(context.Background(), 0, nil, ""))
+			require.NoError(t, buff.Use(tc.buff))
+			buff.DiscardPointsAtOrBefore(tc.timestamp)
+			bufferMembersEquals(t, tc.expected, buff)
+		})
+	}
+}
+
+func TestFPointRingBufferRemoveLast(t *testing.T) {
```

Comment:
[nit] We generally follow this convention (albeit not consistently)

```suggestion
func TestFPointRingBuffer_RemoveLast(t *testing.T) {
```

(same feedback applies to other tests)